frontend-react/docs/best-practices.md

```diff
@@ -0,0 +1,443 @@
+# General Best Practices in ReportStream Front-End App
+
+This document serves as a living document for best practices for contributing to our front-end code and processes.  This is not intended as a resource for best practices in JavaScript, React, CSS, or any other core technology we use (there are much better guides for that linked below); rather, it's meant as a guideline on how we use them in this codebase.
+
+Keep in mind that with any other guidelines, there are always going to be exceptions.  In these cases, use your best judgment, talk to your teammates, and make sure to note historical reasoning via inline or review comments.
+
+## JavaScript/TypeScript
+
+### Naming
+
+- Variables and non-component functions (including hooks) use `camelCase`
+ 
+```js
+let hoursSpentOnHomework = 0;
+
+const minutesPerSecond = 60;
+
+const memoizedCallback = useCallback(...);
+
+function doHomework() { ... }
+
+function useHomeworkDoer() { ... }
+```
+
+- Components and file names use `PascalCase`
+```jsx
+// in ExampleClassComponent.tsx
+class ExampleClassComponent extends Component { ... }
+
+// in ExampleFunctionComponent.tsx
+function ExampleFunctionComponent() { ... }
+```
+
+- Avoid generic names like `a|b|c`, `i|j|k`, `tmp`, `obj` and opt for more descriptive ones like `idx`, `child`, `element`, et cetera
+- Add units when conversions are appropriate: `timeInHours`, `sizeInBytes`, `heightInInches`
+- Use predicate naming for computed boolean values: `isReady`, `shouldDisplayTime`, `willUpdate`
+
+### Prefer using function declarations over function expressions for first-class functions
```

Comment:
This is a proposal from me.  I think the implicit benefits we get from function declarations for first-class functions (and components) are worthwhile, but they're not hard blockers either way.